### ISSUE TITLE
Discussion: renderer process stores are not main process store mere proxies

### ISSUE THREAD

**Issue by eric-burel** [hardchor/electron-redux#64]
Hi,
I'd just like to raise this issue because it has been bugging me for a while. In the readme, you declare that the main process becomes the Single Source of Truth, and renderer mere proxies.

I am not sure about that actually. That would be true if this lib was based on state diffing, but here, as far as I understand, you use _action replaying_ as a sync mechanism.

This is different since your synchronization is actually a loose synchronization. In turn, it has 2 consequences:
- you can't guarantee the actual synchronization between stores when actions are missed (nobody cares because you should not miss actions in a normal electron use case but still, interprocess commmunication may theoritically fail)

- however, thanks to its symetry, your model is far more scalable than a simple main/renderer process. In a big appliction, this allow to split state between the renderer (that purely handles UI) and the main (that handles data), or even to create store for background processes. The only shared part are data related reducers of the main, so that all stores works on the same data.

To sum it up:
- main has its own sagas/observable/thunks/whatever, its own reducers, and shared reducers (e.g to handle loaded data) that correspond to the loosely synched state slice.
- rendrer has its own sagas/observable/thunks, its own reducers, and the reducers it shares with the main.

So the main is not a single source of truth, but simply a redux-aware process that you usually use to handle async data loading.

**Better part:** why use the main? actually we could use a third process, so that both the main process and the renderer process are free to go. 
- renderer 1 has its store + relevant reducers, shared and not shared
- renderer N _same_
- main _same_ (actually in your model the main is just a special process)
- any bg process _same_

This way you can avoid to perform time consuming async action may lead to performance issue if you put them in the main.

I think that if this lib is to evolve, it could be in a direction that helps to spawn stores that communicate altogether, so that we can create a distributed redux-everywhere architecture, a kind of lightweight redux message queue.

**Comment by RXminuS** [comment]
Yeah this confused me at first as well especially since that's what https://github.com/samiskin/redux-electron-store does. I actually prefer this method because it means I can still use Immutable.js for the actual state in my renderers which helps with ensuring performance.

**Comment by hardchor** [comment]
Hey @eric-burel,

Thanks so much for your thoughts, and apologies for replying so late.

> you can't guarantee the actual synchronization between stores when actions are missed (nobody cares because you should not miss actions in a normal electron use case but still, interprocess commmunication may theoritically fail)

You are absolutely correct. In addition, the initial architecture of `electron-redux` did not accommodate for `local` actions (which were added later), getting us somewhat further away from the single-source-of-truth statement.

> I am not sure about that actually. That would be true if this lib was based on state diffing, but here, as far as I understand, you use action replaying as a sync mechanism.

Even with state-diffing, you could lose information through IPC. Last time I checked, https://github.com/samiskin/redux-electron-store uses some sort of state synchronisation and state filtering (as opposed to action replaying), which is why I created this library.

> So the main is not a single source of truth, but simply a redux-aware process that you usually use to handle async data loading.

That is also correct, and should probably get updated in the docs.

> Better part: why use the main? actually we could use a third process, so that both the main process and the renderer process are free to go.

The main process being the central point for all (except for local) actions fits in very nicely with the electron architecture. Renderer processes cannot talk to each other directly, so the main process' store needs to be the central hub for all communication.

> This way you can avoid to perform time consuming async action may lead to performance issue if you put them in the main.

In all honesty, I can't think of a scenario where an action(-producer) or reducer would block the main thread long enough for it to actually matter. Due to the nature of the event loop, async actions are non-blocking, and reducers are simple stateless functions. Adding an extra process (to become what is now the main store) would - imho - add more overhead than the performance penalties you've mentioned. However, if this becomes a problem in your specific use-case, this library should not stop you from working around this, which is leading onto the next point...

> We would not need aliasedAction anymore also, because if only the ith process have the relevant reducers/sagas/whatever, it will be the only one to process it.

Bingo! `aliasedAction` was always a bit of a pain point, and having named stores that you can freely forward to would be a way better alternative. This would, however, require quite an API change. Maybe we should open a new issue with a 2.0 roadmap.

*Long story short*:

I'm more than happy to evolve the library, and this is a great starting point. I'd still prefer for the main store to take up this special role as an "action fan" (or at least some sort of other util that could be decoupled from the store, but lives on the main thread), but I'm open for a discussion around this.

**Comment by eric-burel** [comment]
Soooo I finally have a chance to try this:

- I modified the redux middleware to send only actions with scope 'shared': as a default, actions are local (normal redux behaviour), and only actions marked as shared will reach the main process
- I've created a `createSharedAction` helper

This first step leads to a cleaner architecture.
Second step was fixing the "async" issue: currently when actions are dispatched to the main process, they are also eliminated from the redux pipeline. Then the main process sends it back to the emitter with "local" scope.
What I've done instead: actions are marked with the current web content id. They are also sent to the main process. Then the main process sends the action back to all web contents EXCEPT the one that emitted the event in the first place.

The result is that shared actions behaves like "normal" actions (states update synchronously), their order is respected again.

Code samples:
```
```

**Comment by eric-burel** [comment]
One last thing that gives me a hard time is reloading the page. Since it erases all state in the renderer, it creates some discrepencies with the main state. Maybe I should setup a temporary system to memorize the renderer current state (eg in smth persistent like localStorage) and reload it from there.
